src/mlpack/core/optimizers/adam/nadam_update.hpp

```diff
@@ -0,0 +1,172 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_ADAM_NADAM_UPDATE_HPP
+#define MLPACK_CORE_OPTIMIZERS_ADAM_NADAM_UPDATE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Nadam is an optimizer that combines the Adam and NAG.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @techreport{Dozat2015,
+ *   title       = {Incorporating Nesterov momentum into Adam},
+ *   author      = {Timothy Dozat},
+ *   institution = {Stanford University},
+ *   address     = {Stanford},
+ *   year        = {2015}
+ * }
+ * @endcode
+ */
+class NadamUpdate
+{
+ public:
+  /**
+   * Construct the Nadam update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient
+   * @param scheduleDecay The decay parameter for decay coefficients
+   */
+  NadamUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
+              const double beta2 = 0.99,
+              const double scheduleDecay = 4e-3)
+      :epsilon(epsilon),
+      beta1(beta1),
+      beta2(beta2),
+      scheduleDecay(scheduleDecay),
+      iteration(0),
+      cumBeta1(1)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start 
+   * of the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    v = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for Nadam.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    v *= beta2;
+    v += (1 - beta2) * gradient % gradient;
+
+    double beta1T = beta1 * (1 - (0.5 *
+        std::pow(0.96, iteration * scheduleDecay)));
+
+    double beta1T1 = beta1 * (1 - (0.5 *
+        std::pow(0.96, (iteration + 1) * scheduleDecay)));
+
+    cumBeta1 *= beta1T;
+
+    const double biasCorrection1 = 1.0 - cumBeta1;
+
+    const double biasCorrection2 = 1.0 - std::pow(beta2, iteration);
+
+    const double biasCorrection3 = 1.0 - (cumBeta1 * beta1T1);
+
+    /* Note :- arma::sqrt(v) + epsilon * sqrt(biasCorrection2) is approximated
+     * as arma::sqrt(v) + epsilon
+     */
+    iterate -= (stepSize * (((1 - beta1T) / biasCorrection1) * gradient
```

Comment:
Thanks for the correction, was using comparing against the Adam paper. Also, do you mind to remove the comment above, it's somewhat misleading for Nadam.

Reply:
@zoq But, I have used this approximation. Should I still remove it?

Reply:
We used this comment in the Adam optimizer class to make clear the equation is not directly what they used in the paper (they removed the stability parameter for simplicity). But Nadam does, so I think the comment is misleading, it's exactly what they use in the paper.

Reply:
@zoq No, I actually in that case, they have used arma::sqrt(v + epsilon). But in paper http://cs229.stanford.edu/proj2015/054_report.pdf, they have given the formula like in Adam. Should I change the algo or the comment. Actually I have cited this report along with it.

Reply:
oh, you are right, I need to to get more sleep or some glasses.

Reply:
So, it's just fine as it is.

Reply:
It's fine. Just rest as much as required :)